frostsnap_core/src/device_nonces.rs

```diff
@@ -0,0 +1,379 @@
+use alloc::collections::VecDeque;
+use alloc::vec::Vec;
+use chacha20::{
+    cipher::{KeyIvInit, StreamCipher},
+    ChaCha20,
+};
+use rand_core::RngCore;
+use schnorr_fun::{
+    binonce,
+    frost::{NonceKeyPair, PairedSecretShare, PartySignSession, SignatureShare},
+    fun::prelude::*,
+};
+
+use crate::{
+    nonce_stream::{CoordNonceStreamState, NonceStreamId, NonceStreamSegment},
+    SignSessionId, NONCE_BATCH_SIZE,
+};
+
+type ChaChaSeed = [u8; 32];
+
+#[derive(Clone, Debug, PartialEq, bincode::Encode, bincode::Decode)]
+pub struct SecretNonceSlot {
+    //XXX: index must be first so we can decode it easily without decoding the whole thing
+    //XXX: index == u32::MAX is invalid because we use it to represent "uninitialized"
+    pub index: u32,
+    pub nonce_stream_id: NonceStreamId,
+    pub ratchet_prg_seed: ChaChaSeed,
+    pub signing_state: Option<SigningState>,
+}
+
+#[derive(Clone, Debug, PartialEq, bincode::Encode, bincode::Decode)]
+pub struct SigningState {
+    pub session_id: SignSessionId,
+    pub signature_shares: Vec<SignatureShare>,
+}
+
+pub trait NonceStreamSlot {
+    fn read_index(&mut self) -> Option<u32>;
+    fn read_slot(&mut self) -> Option<SecretNonceSlot>;
+    fn write_slot(&mut self, value: &SecretNonceSlot);
+}
+
+/// Manages two writable sectors of persistent storage such that we make sure the state of the system we're managing is never lost.
+/// The new state is first written, if that succeeds we finally write over the previous state.
+/// In this case we're managing the secret seed needed to produce nonces for signatures.
+#[derive(Clone, Debug, PartialEq)]
+pub struct ABSlot<S> {
+    slots: [S; 2],
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct ABSlots<S> {
+    slots: Vec<ABSlot<S>>,
+}
+
+impl<S: NonceStreamSlot> ABSlots<S> {
+    pub fn new(ab_slots: impl IntoIterator<Item = ABSlot<S>>) -> Self {
+        Self {
+            slots: ab_slots.into_iter().collect(),
+        }
+    }
+
+    pub fn get_or_create(
+        &mut self,
+        stream_id: NonceStreamId,
+        rng: &mut impl RngCore,
+    ) -> &mut ABSlot<S> {
+        let mut i = 0;
+        let found = loop {
+            let ab_slot = &mut self.slots[i];
+            match ab_slot.nonce_stream_id() {
+                Some(slot_nsid) => {
+                    if slot_nsid == stream_id {
+                        break Some(i);
+                    }
+                }
+                None => {
+                    ab_slot.initialize(stream_id, rng);
+                    break Some(i);
+                }
+            }
+            i += 1;
+            if i >= self.slots.len() {
+                break None;
+            }
+        };
+
+        match found {
+            Some(i) => &mut self.slots[i],
+            None => {
+                let random_overwrite_idx = rng.next_u32() as usize % self.slots.len();
```

Comment:
What's the motivation for replacing a random slot?
Prevent attempts of gaming of some kind? Or to spread nonce-slot NOR wear across the sectors? Something else?

An alternative could be to replace the least recently used - but probably much more complex with some way to track.

Imagining if I have some key with devices on coordinator A, and I make a new key with a subset of those devices on a new coordinator B; if those devices have full slots, some of them may opaquely overwrite the nonce-slots belonging to coordinator A.

Then, next time I go to sign with coordinator A, I visit devices one by one, but maybe the last device no longer has knowledge of that nonce stream, and I have to start the signing session again?

Reply:
Good point. I'm going to address this in another PR.